File: RetrofitReservation/app/src/main/java/ma/ensa/retrofitreservation/AddReservationActivity.java
package ma.ensa.retrofitreservation;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.util.Log;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import ma.ensa.retrofitreservation.adapters.ClientSpinnerAdapter;
import ma.ensa.retrofitreservation.models.Chambre;
import ma.ensa.retrofitreservation.models.Client;
import ma.ensa.retrofitreservation.models.Reservation;
import ma.ensa.retrofitreservation.network.RetrofitInstance;
import ma.ensa.retrofitreservation.service.ApiInterface;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AddReservationActivity extends AppCompatActivity {

    Button btnAddReservation;
    TextView startDate, endDate;  // Utilisation de TextView pour afficher la date
    Spinner clientSpinner, chambreSpinner;  // Spinners pour le client et la chambre
    private String Tag = "Reservation";
    private ApiInterface apiInterface;  // Déclaration de l'interface API
    private List<Client> clientsList = new ArrayList<>();  // Liste des clients
    private List<Chambre> chambresList = new ArrayList<>();  // Liste des chambres

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_reservation);

        // Initialisation des vues
        btnAddReservation = findViewById(R.id.btnAddReservation);
        startDate = findViewById(R.id.startDate);
        endDate = findViewById(R.id.endDate);
        clientSpinner = findViewById(R.id.clientSpinner);  // Spinner pour le client
        chambreSpinner = findViewById(R.id.chambreSpinner); // Spinner pour la chambre

        // Récupérer les clients via Retrofit
        getClients();

        // Récupérer les chambres via Retrofit
        getChambres();

        // Action du bouton pour ajouter une réservation
        btnAddReservation.setOnClickListener(v -> createReservation());

        // Ouvrir le DatePicker lorsque l'utilisateur clique sur la date de début
        startDate.setOnClickListener(v -> showDatePickerDialog(startDate));

        // Ouvrir le DatePicker lorsque l'utilisateur clique sur la date de fin
        endDate.setOnClickListener(v -> showDatePickerDialog(endDate));
    }

    private void showDatePickerDialog(TextView dateField) {
        // Obtenir la date actuelle
        Calendar calendar = Calendar.getInstance();
        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DAY_OF_MONTH);

        // Créer un DatePickerDialog
        DatePickerDialog datePickerDialog = new DatePickerDialog(
                this,
                (view, selectedYear, selectedMonth, selectedDayOfMonth) -> {
                    // Afficher la date sélectionnée dans le TextView
                    String selectedDate = selectedDayOfMonth + "/" + (selectedMonth + 1) + "/" + selectedYear;
                    dateField.setText(selectedDate);
                },
                year, month, day);

        // Afficher le DatePickerDialog
        datePickerDialog.show();
    }

    private void createReservation() {
        // Créer un objet Reservation
        Reservation reservation = new Reservation();

        // Récupérer les valeurs des champs
        reservation.setStartDate(startDate.getText().toString());
        reservation.setEndDate(endDate.getText().toString());
        reservation.setClient((Client) clientSpinner.getSelectedItem());
        reservation.setChambre((Chambre) chambreSpinner.getSelectedItem());

        // Envoi de la réservation à l'API
        Call<Reservation> call = RetrofitInstance.getApi().createReservation(reservation);

        call.enqueue(new Callback<Reservation>() {
            @Override
            public void onResponse(Call<Reservation> call, Response<Reservation> response) {
                if (response.isSuccessful()) {
                    Toast.makeText(AddReservationActivity.this, "Réservation ajoutée avec succès", Toast.LENGTH_SHORT).show();
                    Log.d(Tag, response.toString());
                } else {
                    Toast.makeText(AddReservationActivity.this, "Erreur lors de l'ajout de la réservation.", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Reservation> call, Throwable t) {
                Toast.makeText(AddReservationActivity.this, "Erreur : " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void getClients() {
        // Créer un appel Retrofit pour obtenir les clients
        Call<List<Client>> call = RetrofitInstance.getApi().getAllClients();

        call.enqueue(new Callback<List<Client>>() {
            @Override
            public void onResponse(Call<List<Client>> call, Response<List<Client>> response) {
                if (response.isSuccessful()) {
                    clientsList = response.body();

                    // Adapter personnalisé pour le Spinner
                    ClientSpinnerAdapter clientAdapter = new ClientSpinnerAdapter(AddReservationActivity.this, clientsList);
                    clientSpinner.setAdapter(clientAdapter);
                } else {
                    Toast.makeText(AddReservationActivity.this, "Erreur lors de la récupération des clients", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<Client>> call, Throwable t) {
                Toast.makeText(AddReservationActivity.this, "Erreur : " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void getChambres() {
        // Créer un appel Retrofit pour obtenir les chambres
        Call<List<Chambre>> call = RetrofitInstance.getApi().getAllChambres();

        call.enqueue(new Callback<List<Chambre>>() {
            @Override
            public void onResponse(Call<List<Chambre>> call, Response<List<Chambre>> response) {
                if (response.isSuccessful()) {
                    List<Chambre> chambres = response.body();

                    // Remplir le Spinner avec les chambres récupérées
                    ArrayAdapter<Chambre> chambreAdapter = new ArrayAdapter<>(AddReservationActivity.this,
                            android.R.layout.simple_spinner_item, chambres);
                    chambreAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                    chambreSpinner.setAdapter(chambreAdapter);
                } else {
                    Toast.makeText(AddReservationActivity.this, "Erreur lors de la récupération des chambres", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<Chambre>> call, Throwable t) {
                Toast.makeText(AddReservationActivity.this, "Erreur : " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: ReservationRetrofit_KOTLIN/app/src/main/java/ma/ensaj/reservationretrofit_kotlin/AddReservationActivity.kt
package ma.ensaj.reservationretrofit_kotlin

import android.app.DatePickerDialog
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.DatePicker
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import ma.ensaj.reservationretrofit_kotlin.adapters.ClientSpinnerAdapter
import ma.ensaj.reservationretrofit_kotlin.models.Chambre
import ma.ensaj.reservationretrofit_kotlin.models.Client
import ma.ensaj.reservationretrofit_kotlin.models.Reservation
import ma.ensaj.reservationretrofit_kotlin.network.RetrofitInstance
import ma.ensaj.reservationretrofit_kotlin.service.ApiInterface
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.Calendar
import android.view.WindowInsetsController


class AddReservationActivity : AppCompatActivity() {

    private lateinit var btnAddReservation: Button
    private lateinit var startDate: TextView
    private lateinit var endDate: TextView
    private lateinit var clientSpinner: Spinner
    private lateinit var chambreSpinner: Spinner
    private val tag = "Reservation"
    private lateinit var apiInterface: ApiInterface
    private var clientsList: List<Client> = ArrayList()
    private var chambresList: List<Chambre> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_reservation)

        // Enable Edge-to-Edge layout by hiding status and navigation bars
        val windowInsetsController = window.insetsController
        windowInsetsController?.hide(WindowInsetsController.BEHAVIOR_SHOW_BARS_BY_SWIPE)

        // Initialize views
        btnAddReservation = findViewById(R.id.btnAddReservation)
        startDate = findViewById(R.id.startDate)
        endDate = findViewById(R.id.endDate)
        clientSpinner = findViewById(R.id.clientSpinner)
        chambreSpinner = findViewById(R.id.chambreSpinner)

        // Fetch clients and chambres via Retrofit
        getClients()
        getChambres()

        // Button click to create reservation
        btnAddReservation.setOnClickListener { createReservation() }

        // Date Picker for start date
        startDate.setOnClickListener { showDatePickerDialog(startDate) }

        // Date Picker for end date
        endDate.setOnClickListener { showDatePickerDialog(endDate) }
    }

    private fun showDatePickerDialog(dateField: TextView) {
        val calendar = Calendar.getInstance()
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH)
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        // Create a DatePickerDialog
        val datePickerDialog = DatePickerDialog(
            this,
            { _, selectedYear, selectedMonth, selectedDayOfMonth ->
                // Display the selected date in the TextView
                val selectedDate = "$selectedDayOfMonth/${selectedMonth + 1}/$selectedYear"
                dateField.text = selectedDate
            },
            year, month, day
        )

        // Show the DatePickerDialog
        datePickerDialog.show()
    }

    private fun createReservation() {
        // Create a Reservation object
        val reservation = Reservation().apply {
            startDate = this@AddReservationActivity.startDate.text.toString()
            endDate = this@AddReservationActivity.endDate.text.toString()
            client = clientSpinner.selectedItem as Client
            chambre = chambreSpinner.selectedItem as Chambre
        }

        // Send reservation to the API
        val call = RetrofitInstance.getApi().createReservation(reservation)

        call.enqueue(object : Callback<Reservation> {
            override fun onResponse(call: Call<Reservation>, response: Response<Reservation>) {
                if (response.isSuccessful) {
                    Toast.makeText(this@AddReservationActivity, "Réservation ajoutée avec succès", Toast.LENGTH_SHORT).show()
                    Log.d(tag, response.toString())
                } else {
                    Toast.makeText(this@AddReservationActivity, "Erreur lors de l'ajout de la réservation.", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Reservation>, t: Throwable) {
                Toast.makeText(this@AddReservationActivity, "Erreur : ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun getClients() {
        val call = RetrofitInstance.getApi().getAllClients()

        call.enqueue(object : Callback<List<Client>> {
            override fun onResponse(call: Call<List<Client>>, response: Response<List<Client>>) {
                if (response.isSuccessful) {
                    clientsList = response.body() ?: emptyList()

                    // Custom adapter for the client spinner
                    val clientAdapter = ClientSpinnerAdapter(this@AddReservationActivity, clientsList)
                    clientSpinner.adapter = clientAdapter
                } else {
                    Toast.makeText(this@AddReservationActivity, "Erreur lors de la récupération des clients", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<List<Client>>, t: Throwable) {
                Toast.makeText(this@AddReservationActivity, "Erreur : ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun getChambres() {
        val call = RetrofitInstance.getApi().getAllChambres()

        call.enqueue(object : Callback<List<Chambre>> {
            override fun onResponse(call: Call<List<Chambre>>, response: Response<List<Chambre>>) {
                if (response.isSuccessful) {
                    chambresList = response.body() ?: emptyList()

                    // Fill the chambre spinner with retrieved data
                    val chambreAdapter = ArrayAdapter(
                        this@AddReservationActivity,
                        android.R.layout.simple_spinner_item, chambresList
                    )
                    chambreAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                    chambreSpinner.adapter = chambreAdapter
                } else {
                    Toast.makeText(this@AddReservationActivity, "Erreur lors de la récupération des chambres", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<List<Chambre>>, t: Throwable) {
                Toast.makeText(this@AddReservationActivity, "Erreur : ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }
}